### PR TITLE
fix(async-io): instrumented `asyncio.wait_for` properly raises `asyncio.TimeoutError`

### PR DESCRIPTION
**Pull request by smoke**
# Description

As per the example from https://docs.python.org/3.13/library/asyncio-task.html#asyncio.wait_for when `asyncio.wait_for` hits the defined `timeout` it should raise an `asyncio.TimeoutError`,
however this is not the case when `opentelemetry-instrumentation-asyncio` is enabled

If everything is good running the below code in both cases should print `timeout!`

```
# asyncio-wait-for-example.py
import asyncio
import os

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.instrumentation.asyncio import AsyncioInstrumentor

OTEL_PYTHON_DISABLED_INSTRUMENTATIONS = os.getenv("OTEL_PYTHON_DISABLED_INSTRUMENTATIONS", "")
if 'asyncio' not in OTEL_PYTHON_DISABLED_INSTRUMENTATIONS:
    AsyncioInstrumentor().instrument()

trace.set_tracer_provider(TracerProvider())
tracer = trace.get_tracer(__name__)

# code below taken from https://docs.python.org/3.13/library/asyncio-task.html#asyncio.wait_for
async def eternity():
    # Sleep for one hour
    await asyncio.sleep(3600)
    print('yay!')

async def main():
    # Wait for at most 1 second
    try:
        await asyncio.wait_for(eternity(), timeout=1.0)
    except TimeoutError:
        print('timeout!')

asyncio.run(main())

# Expected output:
#
#     timeout!
```

The PR fixes that by not swallowing `asyncio.CancelledError` that is used internally from `asyncio` timeouts.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] disabled asyncio instrumentation `OTEL_PYTHON_DISABLED_INSTRUMENTATIONS="asyncio" python asyncio-wait-for-example.py` prints `timeout!`
- [x] enabled asyncio instrumentation `python asyncio-wait-for-example.py` as well prints `timeout!`

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+- `opentelemetry-instrumentation-aws-lambda` Bugfix: instrumented `asyncio.wait_for` properly raises `asyncio.TimeoutError` as expected https://docs.python.org/3.13/library/asyncio-task.html#asyncio.wait_for
```
Comment (xrmx) [inline CHANGELOG.md:10]:
```suggestion
- `opentelemetry-instrumentation-asyncio` instrumented `asyncio.wait_for` properly raises `asyncio.TimeoutError` as expected
```

**Comment by xrmx** [comment]
@smoke please fix lint

**Comment by smoke** [comment]
> @smoke please fix lint

Done, thanks for noting!

**Review by aabmass — APPROVED** [review]
LGTM but would appreciate a review from @bourbonkk

**Review by xrmx — APPROVED** [review]

**Review by bourbonkk — APPROVED** [review]
LGTM. Thank you for your PR.

**Comment by smoke** [comment]
@xrmx @aabmass I see the branch is out of date, do I need to update it or you will take care?

**Comment by xrmx** [comment]
@smoke we'll take care of it

**Comment by smoke** [comment]
There seems to be a test failing that is not related to this PR https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/9700988497/job/26773822151?pr=2637#step:6:626 

```
=================================== FAILURES ===================================
_______________ TestProgrammatic.test_flask_metrics_new_semconv ________________
/home/runner/work/opentelemetry-python-contrib/opentelemetry-python-contrib/.tox/pypy3-test-instrumentation-flask-1/lib/pypy3.8/site-packages/pytest_benchmark/logger.py:46: PytestBenchmarkWarning: Not saving anything, no benchmarks have been run!
  warner(PytestBenchmarkWarning(text))

self = <tests.test_programmatic.TestProgrammatic testMethod=test_flask_metrics_new_semconv>

    def test_flask_metrics_new_semconv(self):
        start = default_timer()
        self.client.get("/hello/123")
        self.client.get("/hello/321")
        self.client.get("/hello/756")
        duration = max(round((default_timer() - start) * 1000), 0)
        metrics_list = self.memory_metrics_reader.get_metrics_data()
        number_data_point_seen = False
        histogram_data_point_seen = False
        self.assertTrue(len(metrics_list.resource_metrics) != 0)
        for resource_metric in metrics_list.resource_metrics:
            self.assertTrue(len(resource_metric.scope_metrics) != 0)
            for scope_metric in resource_metric.scope_metrics:
                self.assertTrue(len(scope_metric.metrics) != 0)
                for metric in scope_metric.metrics:
                    self.assertIn(metric.name, _expected_metric_names_new)
                    data_points = list(metric.data.data_points)
                    self.assertEqual(len(data_points), 1)
                    for point in data_points:
                        if isinstance(point, HistogramDataPoint):
                            self.assertEqual(point.count, 3)
>                           self.assertAlmostEqual(
                                duration, point.sum, delta=10
                            )
E                           AssertionError: 15 != 0.013619539999979224 within 10 delta (14.9863804[600](https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/9700988497/job/26773822151?pr=2637#step:6:601)0002 difference)
```

**Comment by xrmx** [comment]
@smoke fix for that here https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2645

**Comment by smoke** [comment]
@xrmx I see #2645 is merged and all checks are passing now, I hope to see this PR merged and released sooner than later :raised_hands:

**Pull request merged**